libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -421,6 +436,124 @@ struct ReservedErrorSelector: public PostTypeChecker::Checker
+		))
+			m_willBeWrittenTo = true;
+	}
+private:
+	ASTString const& m_identifierName;
+	bool m_willBeWrittenTo = false;
+};
+
+class LValueChecker: public ASTConstVisitor
+{
+public:
+	LValueChecker(Identifier const& _identifier):
+		m_declaration(_identifier.annotation().referencedDeclaration)
+	{}
+	bool willBeWrittenTo() const { return m_willBeWrittenTo; }
+	void endVisit(Identifier const& _identifier) override
+	{
+		if (m_willBeWrittenTo)
+			return;
+
+		solAssert(_identifier.annotation().referencedDeclaration);
+		if (
+			*_identifier.annotation().referencedDeclaration == *m_declaration &&
+			_identifier.annotation().willBeWrittenTo
+		)
+			m_willBeWrittenTo = true;
+	}
+	void endVisit(InlineAssembly const& _inlineAssembly) override
+	{
+		if (m_willBeWrittenTo)
+			return;
+
+		YulLValueChecker yulChecker{m_declaration->name()};
+		yulChecker(_inlineAssembly.operations());
+		m_willBeWrittenTo = yulChecker.willBeWrittenTo();
+	}
+private:
+	Declaration const* m_declaration{};
+	bool m_willBeWrittenTo = false;
+};
+
+struct SimpleCounterForLoopChecker: public PostTypeChecker::Checker
+{
+	SimpleCounterForLoopChecker(ErrorReporter& _errorReporter): Checker(_errorReporter) {}
+	bool visit(ForStatement const& _forStatement) override
+	{
+		_forStatement.annotation().isSimpleCounterLoop = isSimpleCounterLoop(_forStatement);
+		return true;
+	}
+	bool isSimpleCounterLoop(ForStatement const& _forStatement) const
+	{
+		auto const* simpleCondition = dynamic_cast<BinaryOperation const*>(_forStatement.condition());
+		if (!simpleCondition || simpleCondition->getOperator() != Token::LessThan || simpleCondition->userDefinedFunctionType())
+			return false;
+		if (!_forStatement.loopExpression())
+			return false;
+
+		auto const* simplePostExpression = dynamic_cast<UnaryOperation const*>(&_forStatement.loopExpression()->expression());
+		// This matches both operators ++i and i++
+		if (!simplePostExpression || simplePostExpression->getOperator() != Token::Inc || simplePostExpression->userDefinedFunctionType())
+			return false;
+
+		auto const* lhsIdentifier = dynamic_cast<Identifier const*>(&simpleCondition->leftExpression());
+		auto const* lhsIntegerType = dynamic_cast<IntegerType const*>(simpleCondition->leftExpression().annotation().type);
+		auto const* commonIntegerType = dynamic_cast<IntegerType const*>(simpleCondition->annotation().commonType);
+
+		if (!lhsIdentifier || !lhsIntegerType || !commonIntegerType || *lhsIntegerType != *commonIntegerType)
+			return false;
+
+		if (
+			!dynamic_cast<Identifier const*>(&simpleCondition->rightExpression()) &&
+			!dynamic_cast<Literal const*>(&simpleCondition->rightExpression())
+		)
+		{
+			solAssert(lhsIdentifier);
+			LValueChecker rhsConditionChecker{*lhsIdentifier};
+			simpleCondition->rightExpression().accept(rhsConditionChecker);
+			if (rhsConditionChecker.willBeWrittenTo())
+				return false;
+		}
```

Comment:
All of this can be removed. Instead the ``LValueChecker`` in the very end can just visit the right-hand-side as well, as in:
```
simpleCondition->rightExpression().accept(lhsLValueChecker);
```

(which would allow renaming it back to ``lvalueChecker``)

Reply:
Good point. We can just reuse the same `LValueChecker` instance for both. And the check for `Literal` and `Identifier` is redundant, because those will never modify the variable.